pkg/agent/svid/rotator.go

```diff
@@ -124,15 +136,82 @@ func (r *rotator) SetRotationFinishedHook(f func()) {
 }
 
 func (r *rotator) rotateSVIDIfNeeded(ctx context.Context) (err error) {
-	if rotationutil.ShouldRotateX509(r.clk.Now(), r.state.Value().(State).SVID[0]) {
-		err = r.rotateSVID(ctx)
+	state, ok := r.state.Value().(State)
+	if !ok {
+		return fmt.Errorf("unexpected value type: %T", r.state.Value())
 	}
-	if r.rotationFinishedHook != nil {
-		r.rotationFinishedHook()
+
+	if rotationutil.ShouldRotateX509(r.clk.Now(), state.SVID[0]) {
+		if state.Reattestable {
+			err = r.reattest(ctx)
+		} else {
+			err = r.rotateSVID(ctx)
+		}
+
+		if err == nil && r.rotationFinishedHook != nil {
+			r.rotationFinishedHook()
+		}
 	}
+
 	return err
 }
 
+// reattest goes through the full attestation process with the server and gets a new SVID.
+func (r *rotator) reattest(ctx context.Context) (err error) {
+	counter := telemetry_agent.StartReattestAgentCall(r.c.Metrics)
+	defer counter.Done(&err)
+
+	// Get the mtx before starting the reattestation
+	// In this way, the client do not create new connections until the new SVID is received
+	r.rotMtx.Lock()
+	defer r.rotMtx.Unlock()
+	r.c.Log.Debug("Reattesting node")
+
+	bundle, err := r.getBundle()
+	if err != nil {
+		return err
+	}
+
+	key, err := r.generateKey(ctx)
+	if err != nil {
+		return err
+	}
+
+	csr, err := util.MakeCSRWithoutURISAN(key)
+	if err != nil {
+		return err
+	}
+
+	if r.conn == nil {
```

Comment:
I've been stewing on this for a bit.... I think we should redial here every time instead of caching the connection, for a few reasons.

1. server is going to terminate the subconn for this connection after 3m, requiring the subconn to reconnect. I don't know if the reconnect happens on demand or if it always reestablishes.
2. in either case,  it doesn't make sense to keep a long lived connection open that is only going to be used very occasionally

Reply:
removed reusable connection.